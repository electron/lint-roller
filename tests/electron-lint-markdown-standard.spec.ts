import * as cp from 'node:child_process';
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';

const FIXTURES_DIR = path.resolve(__dirname, 'fixtures');

function runLintMarkdownStandard(...args: string[]) {
  return cp.spawnSync(
    process.execPath,
    [path.resolve(__dirname, '../dist/bin/lint-markdown-standard.js'), ...args],
    { stdio: 'pipe', encoding: 'utf-8' },
  );
}

describe('electron-lint-markdown-standard', () => {
  it('requires --root', () => {
    const { status, stdout } = runLintMarkdownStandard('broken-internal-link.md');

    expect(stdout).toContain('Usage');
    expect(status).toEqual(1);
  });

  it('should run clean when there are no errors', () => {
    const { status } = runLintMarkdownStandard('--root', FIXTURES_DIR, 'clean.md');

    expect(status).toEqual(0);
  });

  it('can ignore a glob', () => {
    const { status } = runLintMarkdownStandard(
      '--root',
      FIXTURES_DIR,
      '--ignore',
      '**/dirty.md',
      '{clean,dirty}.md',
    );

    expect(status).toEqual(0);
  });

  it('can ignore multiple globs', () => {
    const { status } = runLintMarkdownStandard(
      '--root',
      FIXTURES_DIR,
      '--ignore',
      '**/cleanable.md',
      '--ignore',
      '**/dirty.md',
      '*.md',
    );

    expect(status).toEqual(0);
  });

  it('can ignore from a file', () => {
    const { status } = runLintMarkdownStandard(
      '--root',
      FIXTURES_DIR,
      '--ignore-path',
      path.resolve(FIXTURES_DIR, 'ignorepaths'),
      '{clean,cleanable,dirty}.md',
    );

    expect(status).toEqual(0);
  });

  it('can detect errors in code blocks', () => {
    const { status, stdout } = runLintMarkdownStandard('--root', FIXTURES_DIR, 'dirty.md');

    expect(stdout.replace(FIXTURES_DIR, '<root>')).toMatchSnapshot();
    expect(status).toEqual(1);
  });

  it('can fix cleanable errors with --fix option', async () => {
    const tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'lint-roller-'));
    fs.copyFile(path.join(FIXTURES_DIR, 'cleanable.md'), path.join(tmpdir, 'cleanable.md'));

    try {
      const { status, stdout } = runLintMarkdownStandard('--fix', '--root', tmpdir, 'cleanable.md');

      expect(
        await fs.readFile(path.join(tmpdir, 'cleanable.md'), { encoding: 'utf-8' }),
      ).toMatchSnapshot();
      expect(stdout).toContain('File has changed: cleanable.md');
      expect(stdout).toContain('There are 0 errors');
      expect(status).toEqual(0);
    } finally {
      await fs.rm(tmpdir, { recursive: true, force: true });
    }
  });

  it('outputs uncleanable errors with --fix option', async () => {
    const tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'lint-roller-'));
    fs.copyFile(path.join(FIXTURES_DIR, 'dirty.md'), path.join(tmpdir, 'dirty.md'));

    try {
      const { status, stdout } = runLintMarkdownStandard('--fix', '--root', tmpdir, 'dirty.md');

      expect(
        await fs.readFile(path.join(tmpdir, 'dirty.md'), { encoding: 'utf-8' }),
      ).toMatchSnapshot();
      expect(stdout).toContain('File has changed: dirty.md');
      expect(stdout).toContain("Expected '===' and instead saw '=='");
      expect(stdout).toContain('There are 4 errors');
      expect(status).toEqual(1);
    } finally {
      await fs.rm(tmpdir, { recursive: true, force: true });
    }
  });
});
