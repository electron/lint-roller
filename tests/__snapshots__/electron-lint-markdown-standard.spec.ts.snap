// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`electron-lint-markdown-standard can detect errors in code blocks 1`] = `
"
   dirty.md
         6:1:      Unexpected var, use let or const instead.
         6:12:     Extra semicolon.
         20:1:     Unexpected var, use let or const instead.
         20:12:    Extra semicolon.
         21:9:     Expected '===' and instead saw '=='.
         21:27:    Strings must use singlequote.
         21:40:    Extra semicolon.
         27:1:     Unexpected var, use let or const instead.
         27:12:    Extra semicolon.
         28:9:     Expected '===' and instead saw '=='.
         28:27:    Strings must use singlequote.
         28:40:    Extra semicolon.
         33:1:     Code block language identifier should be all lowercase
         34:1:     Unexpected var, use let or const instead.
         34:12:    Extra semicolon.

There are 15 errors in '<root>'
"
`;

exports[`electron-lint-markdown-standard can fix cleanable errors with --fix option 1`] = `
"# The Cleanable Readme

This is a markdown file with some javascript code blocks in it.

\`\`\`js
const foo = 1
\`\`\`

Each block is parsed separately, to avoid linting errors about variable
assignment. Notice that \`var foo\` occurs twice in this markdown file,
but only once in each individual snippet.

The following code block has a few issues:

- semicolons
- type-insensitive equality comparison
- double-quoted string

\`\`\`javascript
const foo = 2
console.log('foo is two')
\`\`\`

Same as before, but with metadata:

\`\`\`javascript title='main.js'
const foo = 2
console.log('foo is two')
\`\`\`

This non-js code block should be ignored by the cleaner and the linter:

\`\`\`sh
echo i am a shell command
\`\`\`

It should allow orphan objects:

\`\`\`js
{ some: 'object' }
\`\`\`

and this wrapping kind too:

\`\`\`js
{
  some: 'object',
  with: 'different whitespace and tabbing'
}
\`\`\`

and arrays:

\`\`\`js
[1, 2, 3]
\`\`\`

and wrapped arrays:

\`\`\`js
[
  4,
  5,
  6
]
\`\`\`
"
`;

exports[`electron-lint-markdown-standard outputs uncleanable errors with --fix option 1`] = `
"# The Dirty Readme

This is a markdown file with some javascript code blocks in it.

\`\`\`js
const foo = 1
\`\`\`

Each block is parsed separately, to avoid linting errors about variable
assignment. Notice that \`var foo\` occurs twice in this markdown file,
but only once in each individual snippet.

The following code block has a few issues:

- semicolons
- type-insensitive equality comparison
- double-quoted string

\`\`\`javascript
const foo = 2
if (foo == 1) console.log('foo is one')
\`\`\`

Same as before, but with metadata:

\`\`\`javascript title='main.js'
const foo = 2
if (foo == 1) console.log('foo is one')
\`\`\`

Blocks should have lowercase language identifiers:

\`\`\`JavaScript
const foo = 3
\`\`\`

This non-js code block should be ignored by the linter:

\`\`\`sh
echo i am a shell command
\`\`\`
"
`;
